crates/ruff_python_formatter/src/comments/placement.rs

```diff
@@ -231,116 +231,116 @@ fn handle_in_between_bodies_own_line_comment<'a>(
+    .next()
+    .is_some()
+    {
+        return CommentPlacement::Default(comment);
+    }
 
-        // it now depends on the indentation level of the comment if it is a leading comment for e.g.
-        // the following `elif` or indeed a trailing comment of the previous body's last statement.
-        let comment_indentation =
-            whitespace::indentation_at_offset(locator, comment.slice().range().start())
-                .map(str::len)
-                .unwrap_or_default();
+    // it now depends on the indentation level of the comment if it is a leading comment for e.g.
+    // the following `elif` or indeed a trailing comment of the previous body's last statement.
+    let comment_indentation =
+        whitespace::indentation_at_offset(locator, comment.slice().range().start())
+            .unwrap_or_default()
+            .len();
 
-        if let Some(preceding_indentation) =
-            whitespace::indentation(locator, &preceding).map(str::len)
-        {
-            return if comment_indentation >= preceding_indentation {
-                // `# comment` has the same or a larger indent than the `pass` statement.
-                // It likely is a trailing comment of the `pass` statement.
-                // ```python
-                // if x == y:
-                //     pass
-                //     # comment
-                // else:
-                //     print("noop")
-                // ```
-                CommentPlacement::trailing(preceding, comment)
+    let Some(preceding_indentation) = whitespace::indentation(locator, &preceding).map(str::len)
+    else {
+        return CommentPlacement::Default(comment);
+    };
+
+    // Compare to the last statement in the body
+    match comment_indentation.cmp(&preceding_indentation) {
+        Ordering::Greater => {
+            // The comment might belong to an arbitrarily deeply nested inner statement
+            // ```python
+            // while True:
+            //     def f_inner():
+            //         pass
+            //         # comment
+            // else:
+            //     print("noop")
+            // ```
+            own_line_comment_after_branch(comment, locator, preceding)
+        }
+        Ordering::Equal => {
+            // The comment belongs to the last statement.
+            // ```python
+            // try:
+            //     pass
+            //     # I'm a trailing comment of the `pass`
+            // except ZeroDivisionError:
+            //     print("nooop")
+            // ```
+            CommentPlacement::trailing(preceding, comment)
+        }
+        Ordering::Less => {
+            // The comment is leading on the following block.
+            // ```python
+            // try:
+            //     pass
+            // # I'm a leading comment of the `except` statement.
+            // except ZeroDivisionError:
+            //     print("nooop")
+            // ```
+            if following.is_alternative_branch_with_node() {
```

Comment:
Nit: I get the `with_node` suffix now. I'm leaning towards keeping this function local to `placement.rs`. It seems rather specific to the needs of the formatter.